docs/MIGRATION_TO_V6.md

```diff
@@ -0,0 +1,205 @@
+
+# Caveats
+
+There are a handful of caveats that should be considered before reading the main document.
+
+* The Ogmigo v6 upgrade adds no new functionality beyond v6 struct support and a compatibility layer that allows both v5 and v6 Ogmigo JSON output to be unmarshalled into v6 structs. If anything wasn’t supported in Ogmigo v5, it won’t be supported in Ogmigo v6.
+* Support for the Byron era is limited in Ogmigo v6, and may not work as expected. In fact, as of this writing (Nov. 2023), critical functionality, such as the _compatibility_ module, assumes Byron _isn’t_ supported. If you must have Byron support, further Ogmigo work will be required.
+* Any attempt to use a v6-enabled Ogmigo library will automatically break the code using Ogmigo. Many fundamental structs (e.g., _Block_) have been altered to assume v6 structs. If somebody absolutely must use v5 structs, they’ll have to import the _v5_ module and use those structs. Even then, it is highly recommended to use the _compatibility_ module whenever possible, thereby assisting in a smooth transition to the default (v6) code.
+
+# Support for v6
+
+As mentioned in the caveats, the v6 structs are now the default in the Ogmigo library. Code will have to be written with this in mind if working directly with v6. The Ogmios migration guide discusses the gist of what changed between v5 and v6. In general, other than changing specific references within the structs, there aren’t major changes required to understand the new structs.
+
+Note that in particular, Ogmigo automatically handles the switch from JSON-WSP to JSON-RPC 2.0. You should just be able to let Ogmigo handle everything automatically, simply pointing it to the appropriate Ogmios endpoint.
+
+## Major Struct Changes
+
+The Ogmios architectural decision doc linked above isn’t fully correct when discussing how various structs have changed. If you’re using v5 structs and want to see where to look in v6 structs for specific data, you should look at the _compatibility_ module and see how v5 is translated to v6.
+
+That said, there are a few general caveats to note.
+
+* A couple of more esoteric v6 struct fields can’t be populated by v5, at least not without adding a lot of code for little gain (e.g., consensus proofs). If there are gaps, get in touch but you may have to fill them yourself.
+* Some data in v5 is encoded in Base64 and switched in v6 to Base16. The _compatibility_ module automatically handles conversion. Still, if your code directly uses these entries, you’ll need to adjust as necessary.
+* Value types don’t always carry over between v5 and v6, and aren’t always properly defined by the API docs. We’ve looked at both real-world data and Ogmios unit tests, and have done the best we can to determine what’s allowed. It is still recommended that you double check any values critical to your functionality. If there are discrepancies, please get in touch.
```

Comment:
I am not sure to understand the last point? What does it mean for value types to not "always carry over" ?

Reply:
This is more about subtle changes that aren't in the doc but will nip you when getting down to the nuts-and-bolts level. Here's a rough list of all the differences I spotted.

- Some values are Base64 in v5 and Base16 in v6. What I found for v6, while probably not exhaustive, is `Block.Issuer.LeaderValue`, `Block.Issuer.OperationalCertificate.Kes.VerificationKey`, `Block.Issuer.VrfVerificationKey`, and `Block.CBOR`.
- The manner in which `Value` is encoded has changed. (This is perhaps the biggest change for us.) It was `Value.Coins` and `Value.Asset`. Now, it's a double-nested BigInt map (i.e., `map[string]map[string]BigInt`. This affects quite a few fields.
- It is noted that `Block.Era` is new. Perhaps this is just a nice-to-have but there's no note regarding how getting that info from v5 data. A workaround is required. (In our case, we check the protocol version.) It's not a big deal.

Also, see elsewhere re: JSON formatting of values like `Tx.TotalCollateral` and `Tx.Fee` having different formatting between v5 and v6. There are more examples. One example is `TxIn`. `TxIn.Index` is exactly the same in v5 and v6. However, in `TxIn.Transaction`, the values are the same but `TxIn.Transaction.Hash` becomes `TxIn.Transaction.ID`. There are several subtle changes like that.